File: C/Tree/Tree_05.c
/*
===============================================================================
Program Name    : Binary Tree - Postorder Traversal
Program Number  : 005
File Name       : Tree_05.c
Description     : This program demonstrates the Postorder Traversal of a
                  Binary Search Tree (BST) using C. The traversal visits
                  the Left Subtree, Right Subtree, and Root Node (LRD)
                  recursively.

Date            : August 2026
Language        : C
Author          : Ishwari Vijaykumar Surve
===============================================================================

Concepts Used:
1. Structure
2. Self-Referential Structure
3. Pointer
4. Binary Search Tree (BST)
5. Recursion
6. Tree Traversal

Key Learning Points:
- Understand Postorder Traversal (LRD).
- Learn recursive tree traversal.
- Visit every node exactly once.
- Display the Root after traversing both subtrees.

===============================================================================
*/
#include<stdio.h>
#include<stdlib.h>
#include<stdbool.h>

#pragma pack(1)
struct node
{
    int data;
    struct node *lchild;
    struct node *rchild;
};

typedef struct node NODE;
typedef struct node* PNODE;
typedef struct node** PPNODE;

// L    D   R
void Inorder(PNODE first)
{
    if(first != NULL)
    {
        Inorder(first -> lchild);
        printf("%d\n",first -> data);
        Inorder(first -> rchild);
    }
}

// D    L   R
void Preorder(PNODE first)
{
    if(first != NULL)
    {
        printf("%d\n",first -> data);
        Preorder(first -> lchild);
        Preorder(first -> rchild);
    }
}

// L    R   D
void Postorder(PNODE first)
{
    if(first != NULL)
    {
        Postorder(first -> lchild);
        Postorder(first -> rchild);
        printf("%d\n",first -> data);
    }
}

void Insert(PPNODE first, int iNo)
{
    PNODE newn = NULL;
    PNODE temp = NULL;

    newn = (PNODE)malloc(sizeof(NODE));

    newn -> data = iNo;
    newn -> lchild = NULL;
    newn -> rchild = NULL;

    if(*first == NULL)
    {
        *first = newn;
    }
    else
    {
        temp = *first;

        while(1)
        {
            if(iNo > temp -> data)
            {
                if(temp -> rchild == NULL)
                {
                    temp -> rchild = newn;
                    break;
                }
                temp = temp -> rchild;
            }
            else if(iNo < temp -> data)
            {
                if(temp -> lchild == NULL)
                {
                    temp -> lchild = newn;
                    break;
                }
                temp = temp -> lchild;
            }
            else if(iNo == temp -> data)
            {
                printf("Unable to Insert as Elements is Duplicate");
                free(newn);
                break;
            }
        }
    }
}

int main()
{
    PNODE head = NULL;

    Insert(&head,11);
    Insert(&head,5);
    Insert(&head,17);

    printf("Inorder Display: \n");
    Inorder(head);

    printf("Preorder Display: \n");
    Preorder(head);
    
    printf("Postorder Display: \n");
    Postorder(head);
    return 0;
}
/*
===============================================================================
Expected Output:
===============================================================================

Inorder Display:

5
11
17

Preorder Display:

11
5
17

Postorder Display:

5
17
11

===============================================================================
Memory Representation:
===============================================================================

          11
         /  \
        5    17


Postorder Traversal (LRD)

Step 1 : Visit Left Child  → 5
Step 2 : Visit Right Child → 17
Step 3 : Visit Root        → 11

Output:

5
17
11

===============================================================================
Algorithm:
===============================================================================

1. If the current node is not NULL:
   a. Traverse the left subtree.
   b. Traverse the right subtree.
   c. Display the current node.
2. Stop when NULL is reached.

===============================================================================
Program Flow:
===============================================================================

            Start
               │
               ▼
     Is Current Node NULL?
        │             │
      Yes            No
       │              │
       ▼              ▼
     Return      Traverse Left
                     │
                     ▼
              Traverse Right
                     │
                     ▼
              Display Root
                     │
                     ▼
                    Stop

===============================================================================
Time Complexity:
===============================================================================

Postorder Traversal : O(n)

Space Complexity : O(h)

where,

n = Number of Nodes
h = Height of the Tree

===============================================================================
Conclusion:
===============================================================================

Program 005 demonstrates the Postorder Traversal of a Binary Search Tree
using recursion. The traversal follows the Left-Right-Data (LRD)
sequence, visiting the root node only after traversing both the left and
right subtrees.

===============================================================================
*/

File: Java/Tree/Tree_05.java
/* 
=============================================================================== 
Program Name    : Binary Search Tree - Postorder Traversal 
Program Number  : 05 
File Name       : Tree_05.java 
 
Description     : This program demonstrates Postorder Traversal of a Binary 
                  Search Tree using Java. It first creates a Binary Search 
                  Tree by inserting elements and then displays the elements 
                  using Postorder Traversal. Postorder Traversal visits the 
                  left subtree, right subtree and current node. 
 
Date            : August 2026 
Language        : Java 
Author          : Ishwari Vijaykumar Surve 
=============================================================================== 
 
Concepts Used: 
1. Class 
2. Object-Oriented Programming (OOP) 
3. Self-Referential Class 
4. Constructor 
5. Binary Search Tree (BST) 
6. Dynamic Memory Allocation 
7. Recursion 
8. Insertion Operation 
9. Postorder Traversal 
10. Function Overloading 
 
Key Learning Points: 
- Understand Postorder Traversal of a Binary Search Tree. 
- Learn how recursion is used to traverse a Tree. 
- Visit the left subtree first. 
- Visit the right subtree after the left subtree. 
- Visit the root node after both subtrees. 
- Understand function overloading using two Postorder functions. 
 
=============================================================================== 
*/ 
 
class node 
{ 
    public int data; 
    public node lchild; 
    public node rchild; 
} 
 
class Tree 
{ 
    private node first; 
 
    public Tree() 
    { 
        first = null; 
    } 
 
    /////////////////////////////////////////////////////////////////////////// 
    // 
    // Function Name : Insert 
    // Description   : Insert element into Binary Search Tree 
    // 
    /////////////////////////////////////////////////////////////////////////// 
 
    public void Insert(int iNo) 
    { 
        node newn = null; 
        node temp = null; 
 
        newn = new node(); 
 
        newn.data = iNo; 
        newn.lchild = null; 
        newn.rchild = null; 
 
        if(first == null) 
        { 
            first = newn; 
        } 
        else 
        { 
            temp = first; 
 
            while(true) 
            { 
                if(iNo > temp.data) 
                { 
                    if(temp.rchild == null) 
                    { 
                        temp.rchild = newn; 
                        break; 
                    } 
 
                    temp = temp.rchild; 
                } 
                else if(iNo < temp.data) 
                { 
                    if(temp.lchild == null) 
                    { 
                        temp.lchild = newn; 
                        break; 
                    } 
 
                    temp = temp.lchild; 
                } 
                else 
                { 
                    System.out.println("Duplicate element not allowed"); 
                    break; 
                } 
            } 
        } 
    } 
 
    /////////////////////////////////////////////////////////////////////////// 
    // 
    // Function Name : Postorder 
    // Description   : Display tree using Postorder Traversal 
    // 
    /////////////////////////////////////////////////////////////////////////// 
 
    public void Postorder(node temp) 
    { 
        if(temp != null) 
        { 
            Postorder(temp.lchild); 
 
            Postorder(temp.rchild); 
 
            System.out.println(temp.data); 
        } 
    } 
 
    /////////////////////////////////////////////////////////////////////////// 
    // 
    // Function Name : Postorder 
    // Description   : Start Postorder Traversal from first node 
    // 
    /////////////////////////////////////////////////////////////////////////// 
 
    public void Postorder() 
    { 
        Postorder(first); 
    } 
} 
 
class Tree_05 
{ 
    public static void main(String A[]) 
    { 
        Tree tobj = new Tree(); 
 
        tobj.Insert(11); 
        tobj.Insert(5); 
        tobj.Insert(17); 
        tobj.Insert(4); 
        tobj.Insert(7); 
        tobj.Insert(15); 
        tobj.Insert(21); 
 
        System.out.println("Postorder Traversal : "); 
 
        tobj.Postorder(); 
    } 
} 
 
/* 
=============================================================================== 
Expected Output: 
=============================================================================== 
 
Postorder Traversal : 
4 
7 
5 
15 
21 
17 
11 
 
=============================================================================== 
Memory Representation: 
=============================================================================== 
 
                         11 
                       /    \ 
                      5      17 
                     / \    /  \ 
                    4   7  15  21 
 
Postorder Traversal: 
 
       Left Subtree  ->  Right Subtree  ->  Root 
 
             4  ->  7  ->  5  ->  15  ->  21  ->  17  ->  11 
 
=============================================================================== 
Postorder Traversal Process: 
=============================================================================== 
 
                         11 
                       /    \ 
                      5      17 
                     / \    /  \ 
                    4   7  15  21 
 
Step 1 : Visit left subtree of 11 
Step 2 : Visit left subtree of 5 
Step 3 : Visit 4 
Step 4 : Visit 7 
Step 5 : Visit 5 
Step 6 : Visit right subtree of 11 
Step 7 : Visit left subtree of 17 
Step 8 : Visit 15 
Step 9 : Visit 21 
Step 10: Visit 17 
Step 11: Visit 11 
 
Final Postorder Sequence: 
 
4  7  5  15  21  17  11 
 
=============================================================================== 
Node Structure: 
=============================================================================== 
 
                    +-----------+-----------+-----------+ 
                    |   data    |  lchild   |  rchild   | 
                    +-----------+-----------+-----------+ 
                          |           |           | 
                        Value       Address     Address 
 
=============================================================================== 
Algorithm: 
=============================================================================== 
 
1. Start the program. 
2. Define the node class with data, lchild and rchild members. 
3. Define the Tree class. 
4. Initialize first to NULL using the constructor. 
5. Define the Insert function. 
6. Insert the given elements into the Binary Search Tree. 
7. Define the recursive Postorder function. 
8. Check whether the current node is NULL. 
9. If the node is not NULL, recursively visit the left subtree. 
10. Recursively visit the right subtree. 
11. Display the data of the current node. 
12. Call the Postorder function starting from the first node. 
13. Display all elements in Postorder sequence. 
14. Stop the program. 
 
=============================================================================== 
Program Flow: 
=============================================================================== 
 
        Start 
          | 
          v 
     Create Tree 
          | 
          v 
   Insert Tree Elements 
          | 
          v 
   Call Postorder() 
          | 
          v 
      first Node 
          | 
          v 
   Visit Left Subtree 
          | 
          v 
   Visit Right Subtree 
          | 
          v 
      Display Root 
          | 
          v 
    Display Elements 
          | 
          v 
         Stop 
 
=============================================================================== 
Time Complexity: 
=============================================================================== 
 
Insertion          : O(log n) Average 
Insertion          : O(n) Worst Case 
Postorder Traversal: O(n) 
 
=============================================================================== 
Space Complexity: 
=============================================================================== 
 
Tree Storage     : O(n) 
Recursion Stack  : O(h) 
 
Where h is the height of the Binary Search Tree. 
 
=============================================================================== 
Conclusion: 
=============================================================================== 
 
Program 05 demonstrates Postorder Traversal of a Binary Search Tree using 
Java. The recursive Postorder function visits the left subtree first, 
followed by the right subtree and finally the current node. This traversal 
is useful when child nodes need to be processed before their parent node. 
 
=============================================================================== 
*/
